### PR TITLE
update to 0.19.0 (rank modulus patch)

### PR DESCRIPTION
**Pull request by grlee77**
This PR continues on from #85, applying the proposed patch from scikit-image/scikit-image#6103

Checklist
* [ ] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [ ] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by hmaarrfk** [comment]
I don't think you added your patch.

**Comment by grlee77** [comment]
> This looks like the right thing, I should close #86. Or we need to decide which one should be merged into the other. It might be easier to simply add a call to rm_pyx_assoc_c_cpp.sh in the build script on this PR.

Yeah, I can squash a lot of the existing commits and add it here.

**Comment by grlee77** [comment]
I'm not sure how to efficiently remove the cython-generated files in a Windows batch script so I skipped that for now. The intention for 0.19.1 is to not have these files in the source tarball in the first place.

There are at least a couple of actual non-Cython generated `.c` files so we can't just do a global removal of all `.c` files.

**Comment by grlee77** [comment]
@erykoff, if you care about co-author credit and have an email you want to associate with your commits, let me know and I will ammend the commit messages. Thanks for helping out here!

**Comment by erykoff** [comment]
No problem, I don't need any credit.  This is just part of my very long quest to get https://github.com/scikit-image/scikit-image/pull/5249 released... 😄

**Comment by erykoff** [comment]
Oh no what happened to the windows build?  One (new) test failure?

**Comment by grlee77** [comment]
> Oh no what happened to the windows build? One (new) test failure?

I just double checked that I didn't mess anything up in the rebase here. A diff between this branch and #85 shows only the change to the rank filter patch and the build.sh cleanup command (i.e. nothing that should be related to this failure). There was some similar failure on Windows + PyPy in the other branch, but not sure if it ever occurred on the Cpython builds.

**Comment by erykoff** [comment]
I'm pretty sure this is new ... the win64 pypy failure has been there since the recent py310 migration.  Don't know if it's worth restarting the build.  But at least osx_64 is all passing!

**Comment by grlee77** [comment]
There is also a `random_state` optional kwarg to the `skimage.measure.ransac` function used in the failing test. Perhaps we need to also set that in these test cases to make sure we always get a consistent result?

**Comment by erykoff** [comment]
Good news is that the non-pypy windows builds passed!  Bad news is that the travis ppc64le took too long and errored during testing.  Do the tests need to be turned off for those?

**Comment by hmaarrfk** [comment]
I think we can restart teh builds with travis when they fail on master.

I"m not going to be able to debug pypy failures. We could just skip the tests and leave it as a known failure. That way at least people can use part of scikit-image. We could report them in the feedstock/scikit-image but I don't think I want to spend more time than "reporting" them for now.

**Comment by grlee77** [comment]
> I"m not going to be able to debug pypy failures. We could just skip the tests and leave it as a known failure. That way at least people can use part of scikit-image. We could report them in the feedstock/scikit-image but I don't think I want to spend more time than "reporting" them for now.

That sounds fine to me. Do you know the syntax to enable this skip? Can we use a `skip:` key under the tests section like one can do for build? (not sure the name of the appropriate env. var to specify)

**Comment by grlee77** [comment]
I will try using the existing style of skipping just the specific failing cases:
```
{% set tests_to_skip = tests_to_skip + " or test_hough_ellipse_zero_angle or test_hough_ellipse_non_zero" %}  #  [python_impl == 'pypy']
```

File: recipe/meta.yaml [hunk 1]
```diff
@@ -87,7 +87,7 @@ requirements:
 {% set tests_to_skip = tests_to_skip + " or test_apply_parallel_rgb" %}  # [win]
 
 # skip known failure on PyPy
-{% set tests_to_skip = tests_to_skip + " or test_hough_ellipse_zero_angle or test_hough_ellipse_non_zero" %}  #  [python_impl == 'pypy']
+{% set tests_to_skip = tests_to_skip + " or test_hough_ellipse_zero_angle or test_hough_ellipse_non_zero" %}  # [python_impl == 'pypy']
```
Comment (erykoff) [inline recipe/meta.yaml:90]: The test is only failing on windows pypy, right?  So I think this needs an `and win`.

Comment (grlee77) [reply recipe/meta.yaml:90]: Yeah, it was really only one of the PyPy cases. I can make it more selective as you suggest

File: recipe/meta.yaml [hunk 2]
```diff
@@ -81,10 +86,17 @@ requirements:
...
+{% set tests_to_skip = tests_to_skip + " or test_hough_ellipse_zero_angle or test_hough_ellipse_non_zero" %}  # [win and (python_impl == 'pypy')]
+
 test:
   requires:
     - pytest
     - pytest-localserver
+    # 2021/12/04: 0.19.0: hmaarrfk
+    # We noticed that osx had strange failures with cythonized code
+    # Testing with full datasets on OSX to ensure we don't have bugs
+    - pooch  # [osx]
```
Comment (hmaarrfk) [inline recipe/meta.yaml:99]:
```suggestion
    - pooch
```

Can we run a single round of full tests before merging?

**Comment by grlee77** [comment]
Okay, looks like this PyPy test skip worked.

File: recipe/meta.yaml (hunk 2 above)
Comment (grlee77) [reply recipe/meta.yaml:99]: I see that pooch was already listed in `requirements/test.txt` in the main repo, so having this unconditionally seems consistent with that.

File: recipe/meta.yaml [hunk 3]
```diff
@@ -81,10 +86,17 @@ requirements:
...
+# skip known failure on Windows PyPy builds
+{% set tests_to_skip = tests_to_skip + " or test_hough_ellipse_zero_angle or test_hough_ellipse_non_zero" %}  # [win and (python_impl == 'pypy')]
+
 test:
   requires:
     - pytest
     - pytest-localserver
+    # 2021/12/04: 0.19.0: hmaarrfk
+    # We noticed that osx had strange failures with cythonized code
+    # Testing with full datasets on OSX to ensure we don't have bugs
```
Comment (erykoff) [inline recipe/meta.yaml:98]: I think we need to remove this comment, or at least update it.

Comment (grlee77) [reply recipe/meta.yaml:98]: you mean, just remove "on OSX" from the last line?

Comment (erykoff) [reply recipe/meta.yaml:98]: Perhaps; also on the first line.  Maybe just `# Include pooch to ensure full test coverage` or something.

Comment (grlee77) [reply recipe/meta.yaml:98]: Okay, let's just let the current round of tests finish and then we can make that change prior to merging

**Comment by grlee77** [comment]
aside from a couple of timeouts this looks good, so is probably ready to merge
will go ahead and update the comment as suggested above now

**Comment by erykoff** [comment]
A new test failure 😭 
```
    def test_max_batch_size():
        """Small batches are slow, large batches -> large allocations -> also slow.
    
        https://github.com/scikit-image/scikit-image/pull/6035#discussion_r751518691
        """
        coords = np.random.randint(low=0, high=1848, size=(40000, 2))
        tstart = time.time()
        ensure_spacing(coords, spacing=100, min_split_size=50,
                       max_split_size=2000)
        dur1 = time.time() - tstart
    
        tstart = time.time()
        ensure_spacing(coords, spacing=100, min_split_size=50,
                       max_split_size=20000)
        dur2 = time.time() - tstart
    
>       assert dur1 < dur2
E       assert 5.904011011123657 < 5.62450909614563
```
Is this a failure because we're testing something with the CI machines which have variable load, or because there's a random generation without a seed?
And does this need to be patched before merge?

**Comment by grlee77** [comment]
> Is this a failure because we're testing something with the CI machines which have variable load, or because there's a random generation without a seed?
> And does this need to be patched before merge?

I saw this once in CI on the main repo and it passed upon re-running. It is a pretty newly added test case. We should perhaps relax the test condition a bit to account for variable load and architecture differences to avoid such sporadic failures. Mainly we wanted to ensure that the default choice of batch size (2000) is not substantially worse than a much larger default choice.

**Comment by grlee77** [comment]
something like `dur1 < 1.25 * dur2` or `dur1 < 1.5 * dur2`. I will make a PR upstream and we can add another patch here...

**Comment by grlee77** [comment]
I think it is okay to merge with that as-is, though. It is not a bug that would affect users, just an indication that the choice of default batch size may not be computationally optimal.

**Comment by grlee77** [comment]
One failure here that i haven't seen before. Not sure if it is just a fluke

```
viewer\tests\test_widgets.py sssssss                                     [100%]

================================== FAILURES ===================================
_________________________________ test_uint8 __________________________________

    def test_uint8():
>       plt.figure()
        self._tkloaded = False
        # to avoid recursions in the getattr code in case of failure, we
        # ensure that self.tk is always _something_.
        self.tk = None
        if baseName is None:
            import os
            baseName = os.path.basename(sys.argv[0])
            baseName, ext = os.path.splitext(baseName)
            if ext not in ('.py', '.pyc'):
                baseName = baseName + ext
        interactive = 0
>       self.tk = _tkinter.create(screenName, baseName, className, interactive, wantobjects, useTk, sync, use)
E       _tkinter.TclError: Can't find a usable init.tcl in the following directories: 
E           {%PREFIX%\tcl\tcl8.6}
E       
E       D:/bld/scikit-image_1638988643419/_test_env/tcl/tcl8.6/init.tcl: couldn't read file "D:/bld/scikit-image_1638988643419/_test_env/tcl/tcl8.6/init.tcl": No error
E       couldn't read file "D:/bld/scikit-image_1638988643419/_test_env/tcl/tcl8.6/init.tcl": No error
E           while executing
E       "source D:/bld/scikit-image_1638988643419/_test_env/tcl/tcl8.6/init.tcl"
E           ("uplevel" body line 1)
E           invoked from within
E       "uplevel #0 [list source $tclfile]"
E       
E       
E       This probably means that Tcl wasn't installed properly.

..\_test_env\lib\tkinter\__init__.py:2270: TclError
```

**Comment by hmaarrfk** [comment]
@conda-forge-admin please restart cis

**Pull request merged**